streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -687,25 +687,25 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
   /**
    * Executes the provided finalizer after this stream's finalizers run.
    */
-  def ensuring[R1 <: R](fin: ZIO[R1, Nothing, _]): ZStream[R1, E, A] =
+  final def ensuring[R1 <: R](fin: ZIO[R1, Nothing, _]): ZStream[R1, E, A] =
     new ZStream[R1, E, A] {
-      override def process = self.process.ensuring(fin)
+      def process = self.process.ensuring(fin)
```

Comment:
At some point, for stack safety, we should switch over to `class ZStream(val process: ZManaged[...])`.

Reply:
Out of curiosity, would just changing the def to val be enough? Or is the trait to class change necessary as well?

Reply:
Well, you need a way to "unwrap" and then "rewrap", rather than add one new stack frame.

Right now, the outer `process` delegates to the inner `process`, so you end up adding a stack frame for every operation. Eventually you run out of stack.

When you store it as a field that you can extract, map over, and stuff back it, it pushes 100% of the transformation to the ZIO effect type, which is already stack safe.

Reply:
I get that completely. My question was wether this is the same.

`trait Stream { val process: ZManaged[...] }` and `class Stream(val process: ZManaged[...])`.

I always thought the fields in class constructors are only sugar, there is no actual difference.

Reply:
I will address this in a follow up PR.

Reply:
Ah, I see. No, there is no difference, so long as you are using `val` there (the initialization will happen eagerly and will not contribute to stack depth).

Reply:
As per @iravid's request I will address this in another PR and change to class Stream.